third-party/cmake/BuildLuarocks.cmake

```diff
@@ -1,35 +1,89 @@
 
-ExternalProject_Add(luarocks
-  PREFIX ${DEPS_BUILD_DIR}
-  URL ${LUAROCKS_URL}
-  DOWNLOAD_DIR ${DEPS_DOWNLOAD_DIR}/luarocks
-  DOWNLOAD_COMMAND ${CMAKE_COMMAND}
-    -DPREFIX=${DEPS_BUILD_DIR}
-    -DDOWNLOAD_DIR=${DEPS_DOWNLOAD_DIR}/luarocks
-    -DURL=${LUAROCKS_URL}
-    -DEXPECTED_SHA256=${LUAROCKS_SHA256}
-    -DTARGET=luarocks
-    -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
-  BUILD_IN_SOURCE 1
-  CONFIGURE_COMMAND ${DEPS_BUILD_DIR}/src/luarocks/configure
-    --prefix=${DEPS_INSTALL_DIR} --force-config ${LUAROCKS_OPTS}
-    --lua-suffix=jit
-  BUILD_COMMAND ""
-  INSTALL_COMMAND ${MAKE_PRG} bootstrap)
+# BuildLuarocks(CONFIGURE_COMMAND ... BUILD_COMMAND ... INSTALL_COMMAND ...)
+# Reusable function to build luarocks, wraps ExternalProject_Add.
+# Failing to pass a command argument will result in no command being run
+function(BuildLuarocks)
+  cmake_parse_arguments(_luarocks
+    ""
+    ""
+    "CONFIGURE_COMMAND;BUILD_COMMAND;INSTALL_COMMAND"
+    ${ARGN})
 
-list(APPEND THIRD_PARTY_DEPS luarocks)
+  if(NOT _luarocks_CONFIGURE_COMMAND AND NOT _luarocks_BUILD_COMMAND
+        AND NOT _luarocks_INSTALL_COMMAND)
+    message(FATAL_ERROR "Must pass at least one of CONFIGURE_COMMAND, BUILD_COMMAND, INSTALL_COMMAND")
+  endif()
+
+  ExternalProject_Add(luarocks
+    PREFIX ${DEPS_BUILD_DIR}
+    URL ${LUAROCKS_URL}
+    DOWNLOAD_DIR ${DEPS_DOWNLOAD_DIR}/luarocks
+    DOWNLOAD_COMMAND ${CMAKE_COMMAND}
+      -DPREFIX=${DEPS_BUILD_DIR}
+      -DDOWNLOAD_DIR=${DEPS_DOWNLOAD_DIR}/luarocks
+      -DURL=${LUAROCKS_URL}
+      -DEXPECTED_SHA256=${LUAROCKS_SHA256}
+      -DTARGET=luarocks
+      -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
+    BUILD_IN_SOURCE 1
+    CONFIGURE_COMMAND "${_luarocks_CONFIGURE_COMMAND}"
+    BUILD_COMMAND "${_luarocks_BUILD_COMMAND}"
+    INSTALL_COMMAND "${_luarocks_INSTALL_COMMAND}")
+endfunction()
+
+# The luarocks binary location
+set(LUAROCKS_BINARY ${HOSTDEPS_BIN_DIR}/luarocks)
+
+# Arguments for calls to 'luarocks build'
+if(MSVC)
+  # In native Win32 don't pass the compiler/linker to luarocks, the bundled
+  # version already knows, and passing them here breaks the build
+  set(LUAROCKS_BUILDARGS CFLAGS=/MT)
+else()
+  set(LUAROCKS_BUILDARGS CC=${HOSTDEPS_C_COMPILER} LD=${HOSTDEPS_C_COMPILER})
+endif()
+
+if(UNIX OR (MINGW AND CMAKE_CROSSCOMPILING))
+
+  if(USE_BUNDLED_LUAJIT)
+    list(APPEND LUAROCKS_OPTS
+      --with-lua=${HOSTDEPS_INSTALL_DIR}
+      --with-lua-include=${HOSTDEPS_INSTALL_DIR}/include/luajit-2.0)
+  endif()
 
-# The path to the luarocks executable
-set(LUAROCKS_BINARY ${DEPS_BIN_DIR}/luarocks)
-# Common build arguments for luarocks build
-set(LUAROCKS_BUILDARGS CC=${DEPS_C_COMPILER} LD=${DEPS_C_COMPILER})
+  BuildLuarocks(
+    CONFIGURE_COMMAND ${DEPS_BUILD_DIR}/src/luarocks/configure
+      --prefix=${HOSTDEPS_INSTALL_DIR} --force-config ${LUAROCKS_OPTS}
+      --lua-suffix=jit
+    INSTALL_COMMAND ${MAKE_PRG} bootstrap)
+elseif(MSVC)
```

Comment:
Do you mind inserting a blank line above this and removing the blank line below it?  Thanks!